src/camera.rs

```diff
@@ -0,0 +1,162 @@
+use std::cmp::Ordering;
+use std::f32;
+
+use nalgebra::{Matrix4, Point3, Rotation3, Vector3};
+
+const TWO_PI: f32 = f32::consts::PI * 2.0;
+const ZOOM_SPEED_BASE: f32 = 0.95;
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct CameraOptions {
+    pub radius_max: f32,
+    pub radius_min: f32,
+    pub polar_angle_distance_min: f32,
+    pub speed_pan: f32,
+    pub speed_rotate: f32,
+    pub speed_zoom: f32,
+    pub speed_zoom_step: f32,
+    pub fovy: f32,
+    pub znear: f32,
+    pub zfar: f32,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct Camera {
+    aspect_ratio: f32,
+    radius: f32,
+    azimuthal_angle: f32,
+    polar_angle: f32,
+    origin: Point3<f32>,
+    up: Vector3<f32>,
+    options: CameraOptions,
+}
+
+impl Camera {
+    pub fn new(
+        screen_size: [f32; 2],
+        radius: f32,
+        azimuthal_angle: f32,
+        polar_angle: f32,
+        options: CameraOptions,
+    ) -> Camera {
+        Camera {
+            aspect_ratio: screen_size[0] / screen_size[1],
+            radius: clamp(radius, options.radius_min, options.radius_max),
+            azimuthal_angle: fmod(azimuthal_angle, TWO_PI),
+            polar_angle: clamp(
+                polar_angle,
+                options.polar_angle_distance_min,
+                f32::consts::PI - options.polar_angle_distance_min,
+            ),
+            origin: Point3::origin(),
+            up: Vector3::z(),
+            options,
+        }
+    }
+
+    pub fn set_screen_size(&mut self, screen_size: [f32; 2]) {
+        self.aspect_ratio = screen_size[0] / screen_size[1];
+    }
+
+    pub fn reset_origin(&mut self) {
+        self.origin = Point3::origin();
+    }
+
+    pub fn pan_ground(&mut self, dx: f32, dy: f32) {
+        let pan_factor = self.options.speed_pan * self.radius / self.options.radius_max;
+        let ground_translation = Vector3::new(dx, dy, 0.0) * pan_factor;
+        let camera_rotation =
+            Rotation3::new(Vector3::z() * (self.azimuthal_angle - f32::consts::FRAC_PI_2));
+        self.origin += camera_rotation * ground_translation;
+    }
+
+    pub fn pan_screen(&mut self, dx: f32, dy: f32) {
+        let pan_factor = self.options.speed_pan * self.radius / self.options.radius_max;
+        let ground_translation = Vector3::new(dx, dy, 0.0) * pan_factor;
+        let camera_rotation =
+            Rotation3::new(Vector3::z() * (self.azimuthal_angle - f32::consts::FRAC_PI_2));
```

Comment:
Thought about it, but didn't do it since these two occurrences are the only ones, and will likely remain that way. I would definitely extract to a function if there was a third.